### PR TITLE
dbft: untie BlockQueue from worker

### LINKED ISSUE #297: Geth node log contains error `Block found but no relative pending task ...`
@AnnaShaleva @chenquanyu @txhsl When checking geth node log during testing, there are some ERROR message appear, it happens in local private net, devnet, testnet, and mainnet, please check if this is normal:
image

```
...
2024-08-15T05:17:40.942Z	DEBUG	dbft@v0.2.0/dbft.go:223	ignoring old height	{"height": 180432}
2024-08-15T05:17:40.942Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "PrepareResponse", "from": 1, "height": 180432, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:40.942Z	DEBUG	dbft@v0.2.0/dbft.go:223	ignoring old height	{"height": 180432}
2024-08-15T05:17:40.942Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "PrepareResponse", "from": 3, "height": 180432, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:40.942Z	DEBUG	dbft@v0.2.0/dbft.go:223	ignoring old height	{"height": 180432}
2024-08-15T05:17:40.942Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 180432, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:40.942Z	DEBUG	dbft@v0.2.0/dbft.go:223	ignoring old height	{"height": 180432}
2024-08-15T05:17:40.942Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "PrepareResponse", "from": 2, "height": 180432, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:40.942Z	DEBUG	dbft@v0.2.0/dbft.go:223	ignoring old height	{"height": 180432}
2024-08-15T05:17:40.942Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 6, "height": 180432, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:40.942Z	DEBUG	dbft@v0.2.0/dbft.go:223	ignoring old height	{"height": 180432}
INFO [08-15|05:17:44.699] Looking for peers                        peercount=7 tried=376 static=0
2024-08-15T05:17:49.926Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "PrepareRequest", "from": 1, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.926Z	DEBUG	dbft@v0.2.0/dbft.go:322	received empty PrepareRequest
2024-08-15T05:17:49.926Z	INFO	dbft@v0.2.0/dbft.go:329	received PrepareRequest	{"validator": 1, "tx": 0}
2024-08-15T05:17:49.927Z	INFO	dbft@v0.2.0/send.go:97	sending PrepareResponse	{"height": 180433, "view": 0}
2024-08-15T05:17:49.927Z	DEBUG	dbft@v0.2.0/send.go:8	broadcasting message	{"type": "PrepareResponse", "height": 180433, "view": 0}
2024-08-15T05:17:49.928Z	DEBUG	dbft@v0.2.0/check.go:28	check preparations	{"hasReq": true, "count": 2, "M": 5}
2024-08-15T05:17:49.929Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "PrepareResponse", "from": 4, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.929Z	INFO	dbft@v0.2.0/dbft.go:424	received PrepareResponse	{"validator": 4}
2024-08-15T05:17:49.929Z	DEBUG	dbft@v0.2.0/check.go:28	check preparations	{"hasReq": true, "count": 3, "M": 5}
2024-08-15T05:17:49.931Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "PrepareResponse", "from": 4, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.931Z	DEBUG	dbft@v0.2.0/dbft.go:412	ignoring PrepareResponse	{"dup": true, "sor": true, "viewChanging": false, "moreThanF": true}
2024-08-15T05:17:49.931Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "PrepareResponse", "from": 0, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.931Z	INFO	dbft@v0.2.0/dbft.go:424	received PrepareResponse	{"validator": 0}
2024-08-15T05:17:49.932Z	DEBUG	dbft@v0.2.0/check.go:28	check preparations	{"hasReq": true, "count": 4, "M": 5}
2024-08-15T05:17:49.932Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "PrepareResponse", "from": 2, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.932Z	INFO	dbft@v0.2.0/dbft.go:424	received PrepareResponse	{"validator": 2}
2024-08-15T05:17:49.932Z	DEBUG	dbft@v0.2.0/check.go:28	check preparations	{"hasReq": true, "count": 5, "M": 5}
2024-08-15T05:17:49.932Z	INFO	dbft@v0.2.0/send.go:124	sending Commit	{"height": 180433, "view": 0}
2024-08-15T05:17:49.932Z	DEBUG	dbft@v0.2.0/send.go:8	broadcasting message	{"type": "Commit", "height": 180433, "view": 0}
2024-08-15T05:17:49.933Z	DEBUG	dbft@v0.2.0/dbft.go:608	reset timer	{"h": 180433, "v": 0, "delay": "10s"}
2024-08-15T05:17:49.933Z	DEBUG	dbft@v0.2.0/check.go:56	not enough to commit	{"count": 1}
2024-08-15T05:17:49.933Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "PrepareResponse", "from": 6, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.933Z	INFO	dbft@v0.2.0/dbft.go:424	received PrepareResponse	{"validator": 6}
2024-08-15T05:17:49.935Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 1, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.935Z	INFO	dbft@v0.2.0/dbft.go:498	received Commit	{"validator": 1}
2024-08-15T05:17:49.939Z	DEBUG	dbft@v0.2.0/check.go:56	not enough to commit	{"count": 2}
2024-08-15T05:17:49.941Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 0, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.941Z	INFO	dbft@v0.2.0/dbft.go:498	received Commit	{"validator": 0}
2024-08-15T05:17:49.941Z	DEBUG	dbft@v0.2.0/check.go:56	not enough to commit	{"count": 3}
2024-08-15T05:17:49.941Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "PrepareResponse", "from": 6, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.941Z	DEBUG	dbft@v0.2.0/dbft.go:412	ignoring PrepareResponse	{"dup": true, "sor": true, "viewChanging": false, "moreThanF": true}
2024-08-15T05:17:49.941Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "PrepareResponse", "from": 2, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.941Z	DEBUG	dbft@v0.2.0/dbft.go:412	ignoring PrepareResponse	{"dup": true, "sor": true, "viewChanging": false, "moreThanF": true}
2024-08-15T05:17:49.941Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "PrepareResponse", "from": 2, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.941Z	DEBUG	dbft@v0.2.0/dbft.go:412	ignoring PrepareResponse	{"dup": true, "sor": true, "viewChanging": false, "moreThanF": true}
2024-08-15T05:17:49.941Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "PrepareResponse", "from": 3, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.941Z	INFO	dbft@v0.2.0/dbft.go:424	received PrepareResponse	{"validator": 3}
2024-08-15T05:17:49.943Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 6, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.946Z	INFO	dbft@v0.2.0/dbft.go:498	received Commit	{"validator": 6}
2024-08-15T05:17:49.946Z	DEBUG	dbft@v0.2.0/check.go:56	not enough to commit	{"count": 4}
2024-08-15T05:17:49.946Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 4, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.946Z	INFO	dbft@v0.2.0/dbft.go:498	received Commit	{"validator": 4}
2024-08-15T05:17:49.956Z	INFO	dbft@v0.2.0/check.go:65	approving block	{"height": 180433, "hash": "0xe0243d46b177dc82c0ad7602d0faef8cb410b2e96579281d8032dfadc933ad89", "tx_count": 0, "merkle": "0xdd2d52113e2431811adee8ce6556e9e3f5941bb9b469397c1440f3905e6d2b21", "prev": "0x5045b47cd64602890b3083694d2d2a3fb923a455ea6db348e78ffcbe58037460"}
2024-08-15T05:17:49.956Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.956Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 2, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.956Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 6, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.956Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 2, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.956Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 4, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.956Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 2, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.956Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 2, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.956Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 6, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.956Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 4, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.956Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 6, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.956Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 4, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.956Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 2, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
2024-08-15T05:17:49.956Z	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 4, "height": 180433, "view": 0, "my_height": 180433, "my_view": 0}
INFO [08-15|05:17:49.953] Imported new chain segment               number=180,433 hash=36dce9..ab5c11 blocks=1 txs=0 mgas=0.000 elapsed=3.867ms     mgasps=0.000  snapdiffs=1.63MiB    triedirty=0.00B
_**ERROR[08-15|05:17:49.956] Block found but no relative pending task number=180,433 sealhash=e0243d..33ad89 hash=bcf140..d02a5c**_
INFO [08-15|05:17:49.957] New block in the chain                   "dbft index"=180,433 "chain index"=180,433 hash=0x36dce998572da3690a596cf4e47707c664023179cdfaf0c65fa2e9cd85ab5c11 "parent hash"=0x5045b47cd64602890b3083694d2d2a3fb923a455ea6db348e78ffcbe58037460 primary=1 coinbase=0x1212000000000000000000000000000000000003 "mix digest"=0xb745a9fcb79da001f45f13878301b89130349cab2bf8225ffc21a2dfff9d532b
INFO [08-15|05:17:49.957] Fetching latest sealing proposal         "desired number"=180,434
INFO [08-15|05:17:49.957] Commit new sealing work                  number=180,434 sealhash=c3e779..9db0dd "parent hash"=36dce9..ab5c11 etherbase=0x1212000000000000000000000000000000000003 txs=0 gas=0       fees=0          elapsed=1.023ms
INFO [08-15|05:17:49.959] Imported new chain segment               number=180,433 hash=e6042c..eb3106 blocks=1 txs=0 mgas=0.000 elapsed=2.985ms     mgasps=0.000  snapdiffs=1.63MiB    triedirty=0.00B
INFO [08-15|05:17:50.958] Sealing proposal updated                 number=180,434 sealhash=c3e779..9db0dd "parent hash"=36dce9..ab5c11 txs=0
2024-08-15T05:17:50.958Z	INFO	dbft@v0.2.0/dbft.go:107	initializing dbft	{"height": 180434, "view": 0, "index": 5, "role": "Backup"}
...
```

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #297.

@chenquanyu, @txhsl, welcome to review. @songb2, please, test this PR carefully, because a behaviour of block sharing has been changed slightly.

### CODE REVIEW

File: consensus/dbft/blockqueue.go [hunk 1]
```diff
@@ -110,9 +50,11 @@ func (bq *blockQueue) PutBlock(b *types.Block, state *state.StateDB, receipts []
 			return fmt.Errorf("failed to insert block into chain: %w", err)
 		}
 		log.Info("Successfully inserted new block", "number", b.Number(), "hash", hash)
+
```
Comment (roman-khimov) [inline consensus/dbft/blockqueue.go:53]: :eyes:

File: consensus/dbft/blockqueue.go [hunk 2]
```diff
@@ -129,7 +71,7 @@ func (bq *blockQueue) PutBlock(b *types.Block, state *state.StateDB, receipts []
 		logs = append(logs, receipt.Logs...)
 	}
 	// Commit block and state to database.
-	_, err := bq.chain.WriteBlockAndSetHead(b, receipts, logs, state, true)
+	_, err := bq.chain.WriteBlockAndSetHead(b, receipts, logs, state, b.Number().Cmp(currH) > 0)
```
Comment (roman-khimov) [inline consensus/dbft/blockqueue.go:77]: This comparison is inherently racy. But in the worst case it's just a change of head, not likely to be a problem.

**Review by roman-khimov — APPROVED** [review]
Looks like we can internalize miner functionality with 1.14 (that doesn't have miner anyway).

File: consensus/dbft/blockqueue.go (hunk 1 above)
Comment (AnnaShaleva) [reply consensus/dbft/blockqueue.go:53]: Fixed!

File: consensus/dbft/blockqueue.go (hunk 2 above)
Comment (AnnaShaleva) [reply consensus/dbft/blockqueue.go:77]: > But in the worst case it's just a change of head

Head is always set, the last condition is an emission of event, and event is not important for our case.

**Comment by AnnaShaleva** [comment]
> Looks like we can internalize miner functionality

I hope we won't, miner should not exist for dBFT. I know there's a lot of copy-past, but I just don't have other thoughts how to do it properly, hope that #233 will solve this problem.

**Review by roman-khimov — APPROVED** [review]

**Comment by AnnaShaleva** [comment]
Updated onto fresh master, conflicts are resolved.

**Review by roman-khimov — APPROVED** [review]

**Comment by songb2** [comment]
@AnnaShaleva dbft unit test failed 
`# github.com/ethereum/go-ethereum/eth/protocols/dbft [github.com/ethereum/go-ethereum/eth/protocols/dbft.test]
eth/protocols/dbft/handler_test.go:32:33: cannot use func(u uint64, address common.Address) bool {…} (value of type func(u uint64, address "github.com/ethereum/go-ethereum/common".Address) bool) as func(uint64, "github.com/ethereum/go-ethereum/common".Address) error value in argument to New
eth/protocols/dbft/handler_test.go:33:33: cannot use func(u uint64, address common.Address) bool {…} (value of type func(u uint64, address "github.com/ethereum/go-ethereum/common".Address) bool) as func(uint64, "github.com/ethereum/go-ethereum/common".Address) error value in argument to New
eth/protocols/dbft/handler_test.go:34:33: cannot use func(u uint64, address common.Address) bool {…} (value of type func(u uint64, address "github.com/ethereum/go-ethereum/common".Address) bool) as func(uint64, "github.com/ethereum/go-ethereum/common".Address) error value in argument to New
eth/protocols/dbft/pool_test.go:20:15: cannot use bc (variable of type *testChain) as Ledger value in argument to NewPool: *testChain does not implement Ledger (wrong type for method IsAddressAllowed)
                have IsAddressAllowed("github.com/ethereum/go-ethereum/common".Address) bool
                want IsAddressAllowed("github.com/ethereum/go-ethereum/common".Address) error
eth/protocols/dbft/pool_test.go:50:38: cannot use bc (variable of type *testChain) as Ledger value in argument to NewPool: *testChain does not implement Ledger (wrong type for method IsAddressAllowed)
                have IsAddressAllowed("github.com/ethereum/go-ethereum/common".Address) bool
                want IsAddressAllowed("github.com/ethereum/go-ethereum/common".Address) error
eth/protocols/dbft/pool_test.go:53:15: cannot use bc (variable of type *testChain) as Ledger value in argument to NewPool: *testChain does not implement Ledger (wrong type for method IsAddressAllowed)
                have IsAddressAllowed("github.com/ethereum/go-ethereum/common".Address) bool
                want IsAddressAllowed("github.com/ethereum/go-ethereum/common".Address) error
eth/protocols/dbft/pool_test.go:77:15: cannot use bc (variable of type *testChain) as Ledger value in argument to NewPool: *testChain does not implement Ledger (wrong type for method IsAddressAllowed)
                have IsAddressAllowed("github.com/ethereum/go-ethereum/common".Address) bool
                want IsAddressAllowed("github.com/ethereum/go-ethereum/common".Address) error
FAIL    github.com/ethereum/go-ethereum/eth/protocols/dbft [build failed]`

**Comment by AnnaShaleva** [comment]
> dbft unit test failed

@songb2, Nice catch, I forgot to fix them when updated #298. Fixed, please review. Let's also have #305 for further development.

**Comment by songb2** [comment]
Have tested it in my private environment, the logs and nodes works well, no issues found.

**Pull request merged**